src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -195,55 +178,110 @@ private void AnalyzeExpressionStatement(OperationAnalysisContext context)
+                            (assignment.Target.Kind == OperationKind.LocalReference ||
+                             assignment.Target.Kind == OperationKind.ParameterReference))
+                        {
+                            return true;
+                        }
+
+                        // 3. For anonymous functions parented by delegate creation, we analyze the parent operation.
+                        if (operation.Kind == OperationKind.AnonymousFunction &&
+                            operation.Parent is IDelegateCreationOperation)
+                        {
+                            return IsHandledOperationTreeShape(operation.Parent);
+                        }
+
+                        // 4. Otherwise, conservatively consider this as an unhandled delegate escape.
+                        return false;
                     }
                 }
 
-                private void AnalyzeTuple(OperationAnalysisContext operationAnalysisContext)
+                private void AnalyzeLocalOrParameterReference(OperationAnalysisContext operationAnalysisContext)
                 {
-                    var tupleOperation = (ITupleOperation)operationAnalysisContext.Operation;
-                    if (!_delegateWrappedInTuple &&
-                        tupleOperation.Elements.Any(o => o.Type.IsDelegateType()))
+                    if (operationAnalysisContext.Operation is IParameterReferenceOperation parameterReference)
                     {
-                        _delegateWrappedInTuple = true;
+                        _referencedParameters.GetOrAdd(parameterReference.Parameter, true);
                     }
-                }
 
-                private void AnalyzeFieldOrPropertyReference(OperationAnalysisContext operationAnalysisContextContext)
-                {
-                    var fieldOrPropertyReference = operationAnalysisContextContext.Operation;
-                    if (!_delegateAssignedToFieldOrProperty &&
-                        fieldOrPropertyReference.Type.IsDelegateType() &&
-                        fieldOrPropertyReference.Parent is ISimpleAssignmentOperation simpleAssignment &&
-                        simpleAssignment.Target.Equals(fieldOrPropertyReference))
+                    if (!_hasDelegateEscape)
                     {
-                        _delegateAssignedToFieldOrProperty = true;
+                        _hasDelegateEscape = !IsHandledOperationTreeShape(operationAnalysisContext.Operation);
                     }
-                }
 
-                private void AnalyzeParameterReference(OperationAnalysisContext operationAnalysisContextContext)
-                {
-                    var parameter = ((IParameterReferenceOperation)operationAnalysisContextContext.Operation).Parameter;
-                    _referencedParameters.GetOrAdd(parameter, true);
+                    return;
+
+                    // Local functions
+                    bool IsHandledOperationTreeShape(IOperation operation)
```

Comment:
can we explan why we need both of these local functions (that are called the same thing)?  and seem to be checking the same thing?  Just a small intuition about the breakdown of why we need both, and which cases go with wwhich.

Reply:
First local function is checking operation tree shape in context of a delegate creation or anonymous function, while the second one is check the tree shape in presence of a parameter/local reference of a delegate type.

Reply:
Refactored, renamed and added doc comments to methods. Hopefully it is more clear now.